### ISSUE TITLE
Reducing keeper replica size in kubernetes

### ISSUE THREAD

**Issue by gintautassulskus** [sorintlab/stolon#360]
Hi,

My question is related to Stolon deployment on Kubernetes.
I have a stateful set of stolon-keepers with the replication factor of 3. 
When I reduce the number of stolon-keeper replicas to 2, the excess node is terminated. stolonctl reports cluster status correctly as shown below. 'stolon_keeper_2' is the pod which has been terminated.

```
=== Keepers ===

UID              HEALTHY   PG LISTENADDRESS        PG HEALTHY    PG WANTEDGENERATION   PG CURRENTGENERATION
stolon_keeper_0  true    6.14.80.8:5432          true            4                       4
stolon_keeper_1  true    6.12.128.10:5432        true            2                       2
stolon_keeper_2  false   6.15.112.9:5432         false           2                       2
```
When I set stateful set replication number back to 3, Kubernetes creates a pod with a name 'stolon_keeper_2'. The newly added node is not reflected in the cluster status and the old status with the old IP is still shown on the list.
In the logs I see that 'stolon_keeper_2' picks up configuration of the earlier terminated pod with the same name 'stolon_keeper_2'.

Can stolon handle changes of the replication factor?
What is the expected workflow in this situation? Should I remove non-existing node entries from etcd manually before increasing replication number?

Thanks, 
Gin

**Comment by sgotti** [comment]
@gintautassulskus Usually when you scale down a statefulset the related PVC is kept so when scaling up again the previous PVC and related PV are reused and should work since the keeper cluster data and the pv data refers the previous keeper identity.

Is this you case or did you remove the PVC? In this case also the keeper identity in the cluster data should be removed. There's an open issue for this: #87

**Comment by gintautassulskus** [comment]
@sgotti, at the moment I am using emptyDir: {}. My primary concern now is to test the controller (stolon).

What I did was just:
1. Set stolon-keeper replication from 3 to 2
2. Waited a little bit until stolon picked up the changes and set replication back to 3.

The outcome was:
1. No changes in the keeper list, - same old IP address.
2. No postgres directory in keeper-2 pod.

The scaled-up pod returns:
```
2017-10-02T11:04:16.527Z	DEBUG	keeper/keeper.go:932	db status	{"started": false}
2017-10-02T11:04:16.527Z	INFO	keeper/keeper.go:1298	database cluster not initialized
2017-10-02T11:04:16.527Z	DEBUG	keeper/keeper.go:1309	target role	{"targetRole": "standby"}
2017-10-02T11:04:16.527Z	INFO	keeper/keeper.go:1348	our db requested role is standby	{"followedDB": "13797f4e"}
2017-10-02T11:04:16.527Z	INFO	keeper/keeper.go:1457	our db role is none
```

Cluster dump returns (presumably) an old config with 
`ListenAddress: (string) (len=10) "6.15.112.9",`. The IP belongs to the previously terminated instance and therefore no longer exists on the network.

**Comment by sgotti** [comment]
> at the moment I am using emptyDir: {}.

EmptyDir is bad, the keepers must have persistent storage or you'll lose your data: https://github.com/sorintlab/stolon/blob/master/doc/architecture.md#keepers

> The scaled-up pod returns:
> 
> 2017-10-02T11:04:16.527Z	DEBUG	keeper/keeper.go:932	db status	{"started": false}
> 2017-10-02T11:04:16.527Z	INFO	keeper/keeper.go:1298	database cluster not initialized
> 2017-10-02T11:04:16.527Z	DEBUG	keeper/keeper.go:1309	target role	{"targetRole": "standby"}
> 2017-10-02T11:04:16.527Z	INFO	keeper/keeper.go:1348	our db requested role is standby	{"followedDB": "13797f4e"}
> 2017-10-02T11:04:16.527Z	INFO	keeper/keeper.go:1457	our db role is none

That's correct and caused by the use of emptydir for the keeper's data: the clusterdata says the the keeper should be a standby but it has empty data, so there's a mismatch.

**Comment by gintautassulskus** [comment]
But by increasing the replication status I am adding a new keeper which has nothing in common with the previous one with the exception of the name. I expect the new keeper to not to have any data.

I see two cases here: first, when the new keeper has no data whatsoever; second, the new keeper may have some or all of the data.

**Comment by sgotti** [comment]
> But by increasing the replication status I am adding a new keeper which has nothing in common with the previous one with the exception of the name. I expect the new keeper to not to have any data.

The problems is that in the cluster data there's the old keeper information, so you should manually remove it. IMHO, for how stateful sets are made, scaling down is a pain, since every software have it's own way to work. If you want to scale down then there're some steps that you have to do:

* Remove the previous PVC
* Remove the keeper from the clusterdata #87 

Automating this in the stateful set definition is not possible (using a pre stop hook won't work or cause other issue since it's called every time the pod is deleted).

Scaling down the statefulset could also put the cluster in an uncoverable state if the removed keeper is the current master and the other keepers are not in sync for whatever problem.

For doing this in a controlled state there's the need for an `operator` like done in the etcd-operator and other operators that can fine manage the pods life.

May I ask why you want to scale your stolon cluster? stolon purpose is pg high availabilty so there's no usually the need to scale it. Or are you using it also for reading from replicas and you want to scale replicas?

**Comment by gintautassulskus** [comment]
Thanks, @sgotti, removing keeper from clusterdata worked.

At the moment I am just evaluating Stolon for production use. I want to be aware how Stolon responds to various aspects, including Kubernetes scaling, and in case of the issues, how to recover the cluster.

Reading from replicas would be preferable, but is not a priority now.

**Comment by sgotti** [comment]
@gintautassulskus This is addressed in the doc changes in #430. If you have further questions please ask on gitter or the mailing list. Thanks. Closing.

**Comment by gintautassulskus** [comment]
Thank you, @sgotti, much appreciated.
